### ISSUE TITLE
gRPC connection leaks when channel referenced from class-scope / global-scope state

### ISSUE THREAD

**Issue by dhermes** [grpc/grpc#7849]
The error

```
E0823 21:44:05.544057132   16251 network_status_tracker.c:48] Memory leaked as all network endpoints were not shut down
```

can be [consistently triggered](https://gist.github.com/dhermes/3c5ff09c23840a755eb7d4499ee5d2f9) via

```
$ LEAK=False nosetests min_repro.py --nocapture
$ LEAK=True nosetests min_repro.py --nocapture
```

There is also a "full-fledged" version that runs in [`HEAD`](https://github.com/GoogleCloudPlatform/gcloud-python/tree/94c2743358171142cb210d773330d4e54a67ecb0) on `gcloud-python` but `min_repro.py` only has `grpcio` stable (1.0.0) as a dependency.

/cc @nathanielmanistaatgoogle

**Comment by nathanielmanistaatgoogle** [comment]
I'd be keen to know whether this demonstrates the same behavior when run against gRPC master (`1.1-dev` rather than `1.0.x`).

**Comment by dhermes** [comment]
Can I help you and you help me? What are you having issues with running it? Install `grpio` (whichever version you like) and install `nose` and you should be set.

You'll need to replace `PROJECT_ID` and `ACCESS_TOKEN` with a valid project on the cloud console  (with the `datastore` API turned on) and with a valid token.

You can get a token quickly from the `gcloud` CLI by using `gcloud auth login` and then

```python
import oauth2client.client
# Skip the GAE and GCE checks and just go for CLI.
creds = oauth2client.client.GoogleCredentials._implicit_credentials_from_files()
print(creds.get_access_token().access_token)
```

**Comment by nathanielmanistaatgoogle** [comment]
A more minimal demonstration is
```
import grpc


class ClassWithMutableClassScopeState(object):
    channel = None


def test_leak():
    channel = grpc.secure_channel(
        'bigtable.googleapis.com:443', grpc.ssl_channel_credentials())
    ClassWithMutableClassScopeState.channel = channel
    grpc.channel_ready_future(channel).result()
```
. I was surprised by how the problem requires execution by `nosetests`; just running the code with `python` isn't enough. As I suspected, the problem manifests on with `grpcio` 1.0.0 but not with master.

How severely are you affected? How much of a problem would it be to wait for the fix in `grpcio` 1.1?

**Comment by dhermes** [comment]
> by how the problem requires execution by `nosetests`

It doesn't, I just ran it that way since I started out with the system test. I just re-ran without `nosetests`

```
$ LEAK=False python min_repro.py
...
$ LEAK=True python min_repro.py
...
E0829 13:20:23.015526485   20090 network_status_tracker.c:48] Memory leaked as all network endpoints were not shut down
```

I also tested your snippet with just `python` and it leaked.

----

> How severely are you affected?

Not at all, in the sense that I don't have any applications in production that use gRPC. But it's a memory leak.

----

> How much of a problem would it be to wait for the fix in `grpcio` 1.1?

You tell me.

**Comment by sYnfo** [comment]
Just hit this issue while trying to use gRPC in multithreaded code.

**Comment by nathanielmanistaatgoogle** [comment]
@sYnfo: can you tell us more about your use case and how severely you are affected by it? We may wish to backport the fix from `master` to `v1.0.x` if it turns out to be a significant problem for people.

**Comment by sYnfo** [comment]
@nathanielmanistaatgoogle Don't worry about me, I've worked around, just wanted to let you guys know that people do run into this. :) Should you still want to know the use case let me know.

**Comment by dhermes** [comment]
Yeah my original report was "focused" though I [confirmed][1] it wasn't specific to the test runner

[1]: https://github.com/grpc/grpc/issues/7849#issuecomment-243244511

**Comment by nathanielmanistaatgoogle** [comment]
What is a minimal reproduction case of this bug _outside of any testing context_? What's the minimal reproduction of this bug that is just a `foo.py` that demonstrates the bug when `python foo.py` is executed for some `python`? (And what is the `python` involved in that execution?)

**Comment by dhermes** [comment]
@nathanielmanistaatgoogle You already [provided one][1], except you left our the `if __name__ == '__main__'` part. I [added][2] the `__main__` part as well as an env. var. to turn off leaking and [dropped it in a gist][2].

```
$ LEAK=False python leak_example.py
$ LEAK=True python leak_example.py
E1018 09:42:38.499929680   25706 network_status_tracker.c:48] Memory leaked as all network endpoints were not shut down
```

[1]: https://github.com/grpc/grpc/issues/7849#issuecomment-242567436
[2]: https://gist.github.com/dhermes/5c1fbeda25ffc26d31b890a10e551aa2

**Comment by dhermes** [comment]
@nathanielmanistaatgoogle Has the fix for this been released? Is the min repro provided by me (via you) not sufficient?

**Comment by nathanielmanistaatgoogle** [comment]
The fix for this has not yet been released and the min repro is probably sufficient but I haven't yet found time to confirm.

Since we believe this is fixed in gRPC Python 1.1-and-later (by [this change](https://github.com/grpc/grpc/pull/7538) by @thunderboltsid), and since we believe that gRPC Python 1.1 will appear sometime before the end of the year, we haven't had much reason of our own to take the time to look into this. This issue doesn't yet have a priority label on it - while it's not high-priority for us, is it high-priority for you? How hot should we consider this fire to be burning?

**Comment by dhermes** [comment]
> How hot should we consider this fire to be burning?

Very hot? Most users end up seeing this error, and many of them are using `google-cloud-python` or GAX-generated libraries. So they go from not even **knowing** they were using gRPC to finding out that gRPC is leaking memory (and causing program shutdown to hang for 2-5 seconds).

@jonparrott WDYT?

**Comment by soltanmm-google** [comment]
@nathanielmanistaatgoogle Is this a question of whether or not to backport to `v1.0.x`?

**Comment by hxiong388** [comment]
I work on the google-cloud-ruby repo, and we got this same error from using the grpc Ruby gem v1.0.x. Our reproduction steps involves a Rails application with some configuration, so it's easier to supply that offline if needed. I'm wondering if the problem we're seeing in Ruby land is the same one as Python? Also can this potential fix for grpc/python can also be applied to grpc/ruby?

**Comment by nathanielmanistaatgoogle** [comment]
@soltanmm-google: yes (or "yes plus"); I think this is a question of whether or not to backport [pull request 7538](https://github.com/grpc/grpc/pull/7538) to `v1.0.x` _and_ whether or not there will be another `v1.0.x` release before the first `v1.1.x` release appears.

**Comment by nathanielmanistaatgoogle** [comment]
@hxiong388: Yes, I suspect that what you're seeing in Ruby is analogous to the problem tracked in this issue. Analogous but separate though, so please open a separate issue?

**Comment by dhermes** [comment]
@nathanielmanistaatgoogle From my (and @bjwatson and @geigerj) perspective, **any** release will do, so the questions of `1.0` or `1.1` series is irrelevant to that. (Though for the purposes of releasing for you, it may matter.)

**Comment by samskillman** [comment]
FWIW, just ran into this issue testing out the new(ish?) google.cloud python library, and diagnosed it had something to do with GRPC by somewhat randomly setting `export GOOGLE_CLOUD_DISABLE_GRPC=1` and  noticing that the memory leak warning goes away.  Definitely a pain point from a "new user" perspective. On the other hand, pretty excited that grpc is being used underneath.

**Comment by nathanielmanistaatgoogle** [comment]
Having been fixed on `v1.0.x`, released in `1.0.2`, and been upmerged to `master`, yes, this is now to our knowledge fixed everywhere.
